File: packages/python/src/llm_jury/classifiers/huggingface_adapter.py
from __future__ import annotations

import asyncio
from typing import Any

from .base import ClassificationResult, Classifier


def _score_list(output: Any) -> list[dict]:
    """The ``[{"label", "score"}, ...]`` list for one input.

    With ``top_k`` set at construction, the pipeline wraps a single input's
    scores in an outer list; a call-time ``top_k`` returns them unwrapped.
    """
    if isinstance(output, dict):
        return [output]
    if isinstance(output, list) and output and isinstance(output[0], list):
        return output[0]
    return list(output or [])


class HuggingFaceClassifier(Classifier):
    """Local ``transformers`` text-classification pipeline as the primary classifier.

    ``labels`` sets the label list the jury debates over. When it is omitted, the
    label names come from the model's full score list on the first call.
    Inference runs in a worker thread so it does not block the event loop, and
    results report ``cost_usd=0.0`` because no paid API is called.
    """

    def __init__(
        self, model_name: str, device: str = "cpu", labels: list[str] | None = None
    ) -> None:
        try:
            from transformers import pipeline
        except ImportError as exc:  # pragma: no cover
            raise RuntimeError(
                "transformers is required for HuggingFaceClassifier"
            ) from exc

        self.pipe = pipeline(
            "text-classification", model=model_name, device=device, top_k=None
        )
        self.labels = list(labels) if labels else []

    async def classify(self, text: str) -> ClassificationResult:
        output = await asyncio.to_thread(self.pipe, text)
        results = _score_list(output)
        if not results:
            raise RuntimeError("HuggingFace pipeline returned no scores")
        top = max(results, key=lambda x: x["score"])

        if not self.labels and len(results) > 1:
            self.labels = [item["label"] for item in results]

        return ClassificationResult(
            label=top["label"],
            confidence=float(top["score"]),
            raw_output=results,
            cost_usd=0.0,
        )

File: packages/typescript/src/classifiers/huggingFaceAdapter.ts
import type { ClassificationResult, Classifier } from "./base.ts";

export type HuggingFaceLabelScore = {
  label: string;
  score: number;
};

export type HuggingFacePipeline = (
  text: string,
  options?: Record<string, unknown>,
) =>
  | HuggingFaceLabelScore[]
  | Promise<HuggingFaceLabelScore[]>
  | HuggingFaceLabelScore[][]
  | Promise<HuggingFaceLabelScore[][]>;

export type HuggingFaceClassifierOptions = {
  modelName?: string;
  device?: string;
  pipeline?: HuggingFacePipeline;
  /**
   * Labels the jury debates over. When omitted they are taken from the
   * model's full score list on the first call.
   */
  labels?: string[];
};

// transformers.js returns only the top label unless asked for every score.
// `topk` is the @xenova/transformers v2 option and `top_k` the
// @huggingface/transformers v3 one; null means "all labels" in both.
const ALL_SCORES_OPTIONS: Record<string, unknown> = { topk: null, top_k: null };

function scoreList(raw: unknown): HuggingFaceLabelScore[] {
  if (Array.isArray(raw)) {
    return (Array.isArray(raw[0]) ? raw[0] : raw) as HuggingFaceLabelScore[];
  }
  if (raw && typeof raw === "object" && "label" in raw) {
    return [raw as HuggingFaceLabelScore];
  }
  return [];
}

/**
 * Local transformers.js text-classification pipeline as the primary
 * classifier. Results report `costUsd: 0` because no paid API is called.
 */
export class HuggingFaceClassifier implements Classifier {
  public labels: string[];
  private modelName?: string;
  private device: string;
  private pipeline?: HuggingFacePipeline;

  constructor(options: HuggingFaceClassifierOptions = {}) {
    this.labels = options.labels ? [...options.labels] : [];
    this.modelName = options.modelName;
    this.device = options.device ?? "cpu";
    this.pipeline = options.pipeline;
  }

  async classify(text: string): Promise<ClassificationResult> {
    const runner = await this.resolvePipeline();
    const normalized = scoreList(await runner(text, { ...ALL_SCORES_OPTIONS }));
    if (normalized.length === 0) {
      throw new Error("HuggingFace pipeline returned no scores");
    }

    let top = normalized[0]!;
    for (const item of normalized.slice(1)) {
      if (item.score > top.score) {
        top = item;
      }
    }

    // A single score is not the full label distribution (e.g. an injected
    // pipeline that ignores the all-scores options), so it does not set labels.
    if (this.labels.length === 0 && normalized.length > 1) {
      this.labels = normalized.map((item) => item.label);
    }

    return {
      label: top.label,
      confidence: Number(top.score),
      rawOutput: normalized,
      costUsd: 0,
    };
  }

  private async resolvePipeline(): Promise<HuggingFacePipeline> {
    if (this.pipeline) {
      return this.pipeline;
    }

    if (!this.modelName) {
      throw new Error("Provide modelName or pipeline to HuggingFaceClassifier.");
    }

    try {
      const transformers = (await import("@xenova/transformers")) as {
        pipeline: (task: string, modelName: string, options?: Record<string, unknown>) => Promise<HuggingFacePipeline>;
      };
      this.pipeline = await transformers.pipeline("text-classification", this.modelName, {
        device: this.device,
      });
      return this.pipeline;
    } catch (error) {
      throw new Error(
        "Unable to initialize HuggingFace pipeline. Install @xenova/transformers or inject a pipeline.",
        { cause: error as Error },
      );
    }
  }
}
